feat: Picker titles

nik-rev opened the pull request:
Pickers now have titles. This improves user experience

It also gives us flexibility, for when we want to add additional contextual information into the picker. For instance, https://github.com/helix-editor/helix/pull/11285 would benefit from displaying the current directory of the file picker.

![image](https://github.com/user-attachments/assets/00cf2c65-49cf-4d32-a62b-3d68e71592e2)

|picker type|title|
|-------------|--------|
| thread_picker | -|
| dap_launch | -|
| dap_switch_stack_frame | -|
| diag_picker | Diagnostics |
| symbol_picker | Document Symbols |
| lsp_workspace_command | Workspace Symbols| 
| workspace_symbol_picker | Workspace Symbols |
| goto_reference | Goto Reference |
| goto_single_impl | Goto Implementation |
| file_picker | Files |
| command_palette | Commands |
| changed_file_picker | Changed Files |
| jumplist_picker | Jump List |
| buffer_picker | Buffers |
| global_search | Global Search |

Closes https://github.com/helix-editor/helix/pull/12204

jpaju commented on helix-term/src/commands/typed.rs:
```diff
@@ -1342,6 +1342,7 @@ fn lsp_workspace_command(
                         },
                     )];
                     let picker = ui::Picker::new(
+                        Some("Workspace Symbols"),
```
Should we have something like "LSP Commands" instead of "Workspace Symbols" in here?

nik-rev replied:
Agree

the-mikedavis commented on helix-term/src/ui/picker.rs:
```diff
@@ -287,6 +288,7 @@ impl<T: 'static + Send + Sync, D: 'static + Send + Sync> Picker<T, D> {
     }
 
     pub fn new<C, O, F>(
+        title: Option<&str>,
```
Instead of adding a parameter to `Picker::new` you can add a `with_title` function like `with_preview` that takes the `&str` to use. The downside of changing the signature of `Picker::new` is that it affects every picker, even the ones that would set `None`

nik-rev replied:
Ok, I did it

nik-rev commented:
Note: I made it so that the titles are in the center instead of the left. 

To do this I made a `TitlePosition::{Left,Right}` enum. This is how it looks like:

![image](https://github.com/user-attachments/assets/ceb99535-df46-4065-9966-ed1badba6df2)

darshanCommits commented:
personally I think the left looked better, but okay not anything major.

David-Else commented:
I also think the left looked much better, thanks for this great PR!

nik-rev commented:
Ok, since 3 of you said you prefer the old look, I'll revert it!

nik-rev commented:
Just updated this PR to also use the new File Explorer's current working directory, as I mentioned before

![image](https://github.com/user-attachments/assets/14899287-2662-41bf-afd3-c9e750c3cc1e)

nik-rev commented:
I also added diagnostic count as well
![image](https://github.com/user-attachments/assets/afd82c13-e34d-4b5b-8c08-aa0471abd745)

David-Else commented:
I am afraid I don't think showing the working directory and errors in the title bar look good. They seem out of place and cluttered.

- If we are to display the current directory in the explorer it should show immediately, not only when we browse to a directory
- Having the current directory in the title bar seems out of place, it should be displayed at the top of the actual explorer if anywhere and be integrated into it
- The number of errors are shown in the status line, seems redundant to show them also in the picker and makes the picker interface cluttered.

Sorry to be a downer, giving the pickers titles is a great idea, but I think the latest additions need a rethink. Showing the current directory should be part of the explorer not in the title of the picker.

nik-rev commented:
> I am afraid I don't think showing the working directory and errors in the title bar look good. They seem out of place and cluttered.
> 
>     * If we are to display the current directory in the explorer it should show immediately, not only when we browse to a directory
> 
>     * Having the current directory in the title bar seems out of place, it should be displayed at the top of the actual explorer if anywhere and be integrated into it
> 
>     * The number of errors are shown in the status line, seems redundant to show them also in the picker and makes the picker interface cluttered.
> 
> 
> Sorry to be a downer, giving the pickers titles is a great idea, but I think the latest additions need a rethink. Showing the current directory should be part of the explorer not in the title of the picker.

The file explorer actually always shows the current directory *relative to Helix's*. When Helix's working directory is the same as the explorer's working directory, it makes no sense to show a path as it is empty (`""`)

I've been using the File Explorer for several weeks now, and not knowing its current directory is unfortunate. When I added this change, it instantly became easier to know where you are in the filesystem when you navigate.

Next, what exactly do you mean by "it should be displayed at the top of the actual explorer if anywhere and be integrated into it"? It is *already* displayed at the top. 

If you're talking about displaying it at the top of the preview, that is even worse because the picker's current directory *just makes sense* to group with that directory's children.

Now, you've mentioned that the diagnostic count is in the statusline. This isn't true for everyone, not all people use `render_diagnostics`. Even if *everyone* used it, they only show diagnostics for the current file. Seeing how many diagnostics there are in the Document Diagnostics picker and the Workspace Diagnostics is useful. 

I disagree that it looks cluttered, and "out of place", but I understand this part is subjective. I think that information such as how many diagnostics are in the picker, and picker's current working directory is information that is *useful*.

If you're saying it should be positioned elsewhere, then where? Where do you think the current title + the decorations should be? Here's the thing, `Box` already has a `Box::titled` API. It was extremely easy to use that in this PR.

Could you elaborate on what you mean by "Showing the current directory should be part of the explorer"? Because the explorer is actually a picker, we don't really have a lot of control over it. We don't have anything fancy or custom to display custom components, and a PR that implements that would be really out of scope here, and exactly the same reason why previous File Explorer pull requests were denied. Incorporating so many features by re-using the same APIs is a wonderful idea.

David-Else commented:
> The file explorer actually always shows the current directory relative to Helix's. When Helix's working directory is the same as the explorer's working directory, it makes no sense to show a path as it is empty ("")

> I've been using the File Explorer for several weeks now, and not knowing its current directory is unfortunate. When I added this change, it instantly became easier to know where you are in the filesystem when you navigate.

I don't know the technicalities of what is and is not possible right now in the interface, I am just commenting as a user, so please excuse my lack of knowledge of how it all works behind the scenes.

The most important time for the current directory to be show is when you open the explorer so you know where you are in the file system. I understand this is a technical problem at the moment, but until this is fixed I would consider it a blocker.

> Next, what exactly do you mean by "it should be displayed at the top of the actual explorer if anywhere and be integrated into it"? It is already displayed at the top.

I agree that not knowing the current directory is bad, but showing this in the title looks out of place to me. I think it should be roughly where the arrow points, but I don't know how easy this is to do:

![Screenshot from 2025-01-24 11-45-42](https://github.com/user-attachments/assets/ff60fbc6-9f38-4e68-a6a2-7477f7ac4cb3)

Maybe this is impossible now as you explained, but then I think we should wait rather than having a hack.

Regarding the diagnostic errors, we can just agree to disagree on that.

Thanks for all the great work you are doing, I hope I didn't come across as negative in any way.

nik-rev commented:
> > The file explorer actually always shows the current directory relative to Helix's. When Helix's working directory is the same as the explorer's working directory, it makes no sense to show a path as it is empty ("")
> 
> > I've been using the File Explorer for several weeks now, and not knowing its current directory is unfortunate. When I added this change, it instantly became easier to know where you are in the filesystem when you navigate.
> 
> I don't know the technicalities of what is and is not possible right now in the interface, I am just commenting as a user, so please excuse my lack of knowledge of how it all works behind the scenes.
> 
> The most important time for the current directory to be show is when you open the explorer so you know where you are in the file system. I understand this is a technical problem at the moment, but until this is fixed I would consider it a blocker.
> 
> > Next, what exactly do you mean by "it should be displayed at the top of the actual explorer if anywhere and be integrated into it"? It is already displayed at the top.
> 
> I agree that not knowing the current directory is bad, but showing this in the title looks out of place to me. I think it should be roughly where the arrow points, but I don't know how easy this is to do:
> 
> ![Screenshot from 2025-01-24 11-45-42](https://private-user-images.githubusercontent.com/12832280/406423154-ff60fbc6-9f38-4e68-a6a2-7477f7ac4cb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc3MjA3NjEsIm5iZiI6MTczNzcyMDQ2MSwicGF0aCI6Ii8xMjgzMjI4MC80MDY0MjMxNTQtZmY2MGZiYzYtOWYzOC00ZTY4LWE2YTItNzQ3N2Y3YWM0Y2IzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI0VDEyMDc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4MDEyMjEzZmM1YjI1NWJhODg1ZDFlOTgxZGE1MTg1NjcyNjJjMjdhMDBiOTMyZjM3NGZkY2IyZDdiOTYzYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JYQRjsDzK1sDVpI_IGD-cW_NRmLCUEDg6M_7xtwkpY8)
> 
> Maybe this is impossible now as you explained, but then I think we should wait rather than having a hack.
> 
> Regarding the diagnostic errors, we can just agree to disagree on that.
> 
> Thanks for all the great work you are doing, I hope I didn't come across as negative in any way.

Thanks for this discussion, I think it's good to have!

> > The file explorer actually always shows the current directory relative to Helix's. When Helix's working directory is the same as the explorer's working directory, it makes no sense to show a path as it is empty ("")
> 
> > I've been using the File Explorer for several weeks now, and not knowing its current directory is unfortunate. When I added this change, it instantly became easier to know where you are in the filesystem when you navigate.
> 
> I don't know the technicalities of what is and is not possible right now in the interface, I am just commenting as a user, so please excuse my lack of knowledge of how it all works behind the scenes.
> 
> The most important time for the current directory to be show is when you open the explorer so you know where you are in the file system. I understand this is a technical problem at the moment, but until this is fixed I would consider it a blocker.
> 
> > Next, what exactly do you mean by "it should be displayed at the top of the actual explorer if anywhere and be integrated into it"? It is already displayed at the top.
> 
> I agree that not knowing the current directory is bad, but showing this in the title looks out of place to me. I think it should be roughly where the arrow points, but I don't know how easy this is to do:
> 
> ![Screenshot from 2025-01-24 11-45-42](https://private-user-images.githubusercontent.com/12832280/406423154-ff60fbc6-9f38-4e68-a6a2-7477f7ac4cb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc3MjA3NjEsIm5iZiI6MTczNzcyMDQ2MSwicGF0aCI6Ii8xMjgzMjI4MC80MDY0MjMxNTQtZmY2MGZiYzYtOWYzOC00ZTY4LWE2YTItNzQ3N2Y3YWM0Y2IzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI0VDEyMDc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4MDEyMjEzZmM1YjI1NWJhODg1ZDFlOTgxZGE1MTg1NjcyNjJjMjdhMDBiOTMyZjM3NGZkY2IyZDdiOTYzYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JYQRjsDzK1sDVpI_IGD-cW_NRmLCUEDg6M_7xtwkpY8)
> 
> Maybe this is impossible now as you explained, but then I think we should wait rather than having a hack.
> 
> Regarding the diagnostic errors, we can just agree to disagree on that.
> 
> Thanks for all the great work you are doing, I hope I didn't come across as negative in any way.

I think discussions like these are good to have! When a maintainer reviews, they can leave a comment on how "extra information" (such as the amount of diagnostics, or current directory of the picker) would be best handled

David-Else commented:
> I should've been more clear; Try opening the "File Explorer in Current Buffer's Directory" when you are not at the root

Thanks for your explanation, I had not used your PR for long enough to notice that it instantly showed the current directory when you were not on the root.
